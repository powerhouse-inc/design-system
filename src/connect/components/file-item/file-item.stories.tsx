import { Meta, StoryObj } from '@storybook/react';
import { FileItem } from './file-item';

const meta: Meta<typeof FileItem> = {
    title: 'Connect/Components/FileItem',
    component: FileItem,
    argTypes: {
        onClick: { action: 'onClick' },
        onOptionsClick: { action: 'onOptionsClick' },
        title: { control: 'text' },
        subTitle: { control: 'text' },
        onCancelInput: { action: 'onCancelInput' },
        onSubmitInput: { action: 'onSubmitInput' },
        mode: { control: { type: 'select' }, options: ['read', 'write'] },
        item: { control: { type: 'object' } },
        icon: {
            control: { type: 'select' },
            options: ['legal', 'global', 'profile', 'budget', 'template'],
        },
    },
    decorators: [
        Story => (
            <div className="w-[500px] bg-white p-10">
                <Story />
            </div>
        ),
    ],
};

export default meta;

type Story = StoryObj<typeof meta>;

export const ReadMode: Story = {
    args: {
        mode: 'read',
        title: 'Legal Contract #1',
        subTitle:
            'MakerDAO/Ecosystem Actors/Powerhouse/Chronicle Labs/Legal Contract 1',
        icon: 'profile',
        item: {
            id: '1',
            label: 'Test Folder',
            availableOffline: false,
            path: '',
            type: 'FILE',
        },
    },
};

export const WriteMode: Story = {
    ...ReadMode,
    args: {
        ...ReadMode.args,
        mode: 'write',
    },
};

export const NotAllowedToCreateDocuments: Story = {
    ...ReadMode,
    args: {
        ...ReadMode.args,
        isAllowedToCreateDocuments: false,
    },
};
