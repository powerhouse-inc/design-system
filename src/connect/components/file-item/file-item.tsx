import BudgetImg from '@/assets/icons/budget.png';
import GlobalImg from '@/assets/icons/global.png';
import LegalImg from '@/assets/icons/legal.png';
import ProfileImg from '@/assets/icons/profile.png';
import TemplateImg from '@/assets/icons/template.png';
import {
    ConnectDropdownMenu,
    ConnectDropdownMenuProps,
    TreeItem,
    defaultDropdownMenuOptions,
} from '@/connect';
import {
    DivProps,
    Icon,
    TreeViewInput,
    TreeViewInputProps,
    UseDraggableTargetProps,
    useDraggableTarget,
} from '@/powerhouse';
import React, { useRef, useState } from 'react';
import { twMerge } from 'tailwind-merge';
import { SyncStatusIcon } from '../status-icon';

const iconMap = {
    legal: LegalImg,
    global: GlobalImg,
    profile: ProfileImg,
    budget: BudgetImg,
    template: TemplateImg,
};

export type FileItemIconType = keyof typeof iconMap;

export interface FileItemProps
    extends Omit<DivProps, 'onDragStart' | 'onDragEnd'> {
    title: string;
    subTitle?: string;
    mode?: 'write' | 'read';
    icon?: FileItemIconType;
    customIcon?: React.ReactNode;
    itemOptions?: ConnectDropdownMenuProps['items'];
    isAllowedToCreateDocuments?: boolean;
    onOptionsClick: ConnectDropdownMenuProps['onItemClick'];
    onSubmitInput?: TreeViewInputProps['onSubmitInput'];
    onCancelInput?: TreeViewInputProps['onCancelInput'];
    item: TreeItem;
    onDragStart?: UseDraggableTargetProps<TreeItem>['onDragStart'];
    onDragEnd?: UseDraggableTargetProps<TreeItem>['onDragEnd'];
    displaySyncIcon?: boolean;
}

export const FileItem: React.FC<FileItemProps> = ({
    title,
    subTitle,
    customIcon,
    itemOptions,
    onOptionsClick,
    mode = 'read',
    icon = 'global',
    item,
    isAllowedToCreateDocuments = true,
    onDragEnd,
    onDragStart,
    onCancelInput = () => {},
    onSubmitInput = () => {},
    displaySyncIcon = false,
    ...divProps
}) => {
    const containerRef = useRef(null);
    const [isDropdownMenuOpen, setIsDropdownMenuOpen] = useState(false);

    const { dragProps } = useDraggableTarget({
        data: item,
        onDragEnd,
        onDragStart,
    });

    const isReadMode = mode === 'read';

    const containerStyles = twMerge(
        'group flex h-12 cursor-pointer select-none items-center rounded-lg bg-gray-200 px-2 text-gray-600 hover:text-gray-800',
        divProps.className,
    );

    const iconNode = customIcon || (
        <div className="relative">
            <img
                src={iconMap[icon]}
                alt="file icon"
                className="max-w-none"
                width={32}
                height={34}
            />
            {isReadMode && displaySyncIcon && item.syncStatus && (
                <div className="absolute bottom-[-2px] right-0 size-3 rounded-full bg-white">
                    <div className="absolute left-[-2px] top-[-2px]">
                        <SyncStatusIcon syncStatus={item.syncStatus} />
                    </div>
                </div>
            )}
        </div>
    );

    const content = isReadMode ? (
        <>
            <div className="max-h-6 truncate text-sm font-medium group-hover:text-gray-800">
                {title}
            </div>
            <div className="max-h-6 truncate text-xs font-medium text-gray-600 group-hover:text-gray-800">
                {subTitle}
            </div>
        </>
    ) : (
        <TreeViewInput
            className="ml-3 flex-1 font-medium"
            defaultValue={title}
            onCancelInput={onCancelInput}
            onSubmitInput={onSubmitInput}
        />
    );

    return (
        <div className="relative" ref={containerRef}>
            <div {...dragProps} {...divProps} className={containerStyles}>
                <div className="relative flex flex-1 flex-row items-center overflow-hidden">
                    <div className="mr-1.5">{iconNode}</div>
                    <div
                        className={twMerge(
                            'overflow-hidden text-gray-800',
                            !isReadMode && 'w-full',
                        )}
                    >
                        {content}
                    </div>
                </div>
                {isReadMode && isAllowedToCreateDocuments && (
                    <div
                        onClick={e => {
                            e.stopPropagation();
                            setIsDropdownMenuOpen(true);
                        }}
                    >
                        <Icon
                            name="vertical-dots"
                            className="hidden group-hover:inline-block"
                            size={24}
                        />
                    </div>
                )}
            </div>
            {isAllowedToCreateDocuments && (
                <ConnectDropdownMenu
                    isOpen={isDropdownMenuOpen}
                    onOpenChange={() =>
                        setIsDropdownMenuOpen(!isDropdownMenuOpen)
                    }
                    items={itemOptions || [...defaultDropdownMenuOptions]}
                    menuClassName="bg-white cursor-pointer"
                    menuItemClassName="hover:bg-slate-50 px-2"
                    onItemClick={onOptionsClick}
                    popoverProps={{
                        triggerRef: containerRef,
                        placement: 'bottom end',
                        offset: -10,
                    }}
                />
            )}
        </div>
    );
};
