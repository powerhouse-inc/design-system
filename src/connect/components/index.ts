export * from './drive-view';
export * from './dropdown-menu';
export * from './sidebar';
export * from './tree-view';
export * from './tree-view-item';
