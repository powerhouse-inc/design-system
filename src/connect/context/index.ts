export * from './ItemsContext';
export * from './WagmiContext';
