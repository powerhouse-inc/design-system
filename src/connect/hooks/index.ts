export * from './tree-view';
export * from './useENSInfo';
