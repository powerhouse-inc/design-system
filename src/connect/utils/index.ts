export * from './address';
export * from './get-folder-status';
export * from './mocks';
export * from './node-options';
