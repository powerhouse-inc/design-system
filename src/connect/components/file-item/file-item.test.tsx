import { TreeItem } from '@/connect';
import { fireEvent, render, screen } from '@testing-library/react';
import { FileItem } from './file-item';

const item: TreeItem = {
    id: '1',
    label: 'Test Folder',
    availableOffline: false,
    path: '',
    type: 'FILE',
};

describe('FileItem Component', () => {
    it('should match snapshot', () => {
        const { asFragment } = render(
            <FileItem
                title="Legal Contract #1"
                subTitle="MakerDAO/Ecosystem Actors/Powerhouse/Chronicle Labs/Legal Contract 1"
                icon="profile"
                onClick={() => {}}
                onOptionsClick={() => {}}
                item={item}
            />,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('should call onClick handler', () => {
        const onClick = vi.fn();

        render(
            <FileItem
                title="Legal Contract #1"
                subTitle="MakerDAO/Ecosystem Actors/Powerhouse/Chronicle Labs/Legal Contract 1"
                icon="profile"
                onClick={onClick}
                onOptionsClick={() => {}}
                item={item}
            />,
        );

        fireEvent.click(screen.getByText('Legal Contract #1'));
        expect(onClick).toBeCalledTimes(1);
    });
});
