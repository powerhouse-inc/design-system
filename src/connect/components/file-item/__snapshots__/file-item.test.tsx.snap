// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileItem Component > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="group flex h-12 cursor-pointer select-none items-center rounded-lg bg-gray-200 px-2 text-gray-600 hover:text-gray-800"
  >
    <div
      class="relative flex flex-1 flex-row items-center overflow-hidden"
    >
      <div
        class="mr-1.5"
      >
        <img
          alt="file icon"
          class="max-w-none"
          height="34"
          src="/src/assets/icons/profile.png"
          width="32"
        />
      </div>
      <div
        class="overflow-hidden text-gray-800"
      >
        <div
          class="max-h-6 truncate text-sm font-medium"
        >
          Legal Contract #1
        </div>
        <div
          class="max-h-6 truncate text-xs font-medium text-gray-600"
        >
          MakerDAO/Ecosystem Actors/Powerhouse/Chronicle Labs/Legal Contract 1
        </div>
      </div>
    </div>
    <div>
      <svg
        class="hidden group-hover:inline-block"
        style="width: 24px; height: 24px;"
      >
        <use
          href="icons.svg#vertical-dots"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
